lib/common.js

```diff
@@ -66,6 +68,60 @@ const rsplit = (string, separator) => {
   return [string.slice(0, lastIndex), string.slice(lastIndex + 1)];
 };
 
+class CryptoRandomPrefetcher {
+  constructor(bufSize, valueSize) {
+    this.buf = crypto.randomBytes(bufSize);
+    this.pos = 0;
+    this.vsz = valueSize;
+  }
+
+  next(
+    // Returns Buffer with next `valueSize` random bytes.
+  ) {
+    if (this.pos === this.buf.length) {
+      this.pos = 0;
+      crypto.randomFillSync(this.buf);
+    }
+    const end = this.pos + this.vsz;
+    const buf = this.buf.slice(this.pos, end);
+    this.pos = end;
+    return buf;
+  }
+}
+
+const randPrefetcher = new CryptoRandomPrefetcher(4096, 4);
+const UINT32_MAX = 0xFFFFFFFF;
+const cryptoRandom = () =>
+  randPrefetcher.next().readUInt32LE(0, true) / (UINT32_MAX + 1);
+
+class ExpiringMap extends Map {
+  constructor(interval, ...args) {
+    super(...args);
+
+    this._interval = interval;
+    this._timer = setInterval(() => this._purge(), interval).unref();
+    this._deletionTimes = new Map();
+  }
+
+  _purge() {
+    const purgeTime = Date.now();
+    this._deletionTimes.forEach((delTime, key) => {
+      if (purgeTime - delTime > this._interval) {
+        super.delete(key);
+      }
+    });
+  }
+
+  get(key) {
+    this._deletionTimes.delete(key);
+    return super.get(key);
+  }
+
+  delete(key) {
+    this._deletionTimes.set(key, Date.now());
```

Comment:
This class was created for internal usage, where it is not supposed to call `delete` on the same key multiple times in a row without calling `get` in between those calls to make sure, that the required item is still there, thus resetting the deletion time.

Reply:
Well, even if it's internal we still have to know how it works and this is clearly not an obvious behavior. In the end it's in lib/common which means it may be used later for other things and lead to strange behavior.

Reply:
It is expected to work like that by the code that is using it internally and it is covered by tests, so I can't say that this behavior is not obvious. Also, it is not exported outside, and I don't think it may be used anywhere else since it's usage is very specific in its nature.
Regarding the placement of this class in this file, do you think it will be better to place this class in `lib/server.js` instead?

Reply:
@lundibundi I've provided documentation to all of the new methods and classes in this file, including the `ExpiringMap`, it should be clear now, take a look.